### ISSUE TITLE
No module named 'matplotlib'

### ISSUE THREAD

**Issue by bbigras** [tweag/jupyterWith#176]
**Describe the bug**

**To Reproduce**

```python
import matplotlib.pyplot as plt
```

```nix
let
  jupyter = import (builtins.fetchGit {
    url = https://github.com/tweag/jupyterWith;
    rev = "e80d490476856d36405950fb64cf393d599fe246";
  }) {};

  iPython = jupyter.kernels.iPythonWith {
    name = "python";
    packages = p: with p; [
      matplotlib
    ];
  };

  jupyterEnvironment =
    jupyter.jupyterlabWith {
      kernels = [ iPython ];
    };
in
  jupyterEnvironment.env

```

**Expected behavior**

**Environment**
 - OS name + version: nixos-unstable
 - Version of the code: e80d490476856d36405950fb64cf393d599fe246

**Additional context**

**Comment by guaraqe** [comment]
There are two Python kernels when you start JupyterLab - the default one, empty, and which we cannot remove, and the one you defined with JupyterWith. Can you confirm you picked the correct one?

**Comment by nfrisby** [comment]
I ran into something similar a few months ago. Here are my notes from how I got past it. HTH.

-----

```
$ cat shell.nix
let
  jupyter = import (builtins.fetchGit {
    url = https://github.com/tweag/jupyterWith;
    rev = "";
  }) {};

  iPython = jupyter.kernels.iPythonWith {
    name = "python";
    packages = p: with p; [ scipy matplotlib ipympl ];
  };

  jupyterEnvironment =
    jupyter.jupyterlabWith {
      kernels = [ iPython ];
    };
in
  jupyterEnvironment.env
```

OK. I have it all working. The process is as follows, mainly following the jupyterWith README instructions.
* Start with the `shell.nix` above
* Run `generate-directory @jupyter-widgets/jupyterlab-manager jupyter-matplotlib` in the shell.
* Leave, add the `directory = ./jupyterlab`; line, re-enter.
* Run `jupyter lab`.
* Use the kernel named `Python3 - python`, not the kernel named `Python 3`.

Now everything should work EXCEPT you may need to restart the kernel upon opening your notebook. It seems like its first > attempt to run upon opening (... maybe this is due to the `.ipynb_checkpoints` directory?) fails to load the widgets. But > restarting and rerunning seems to work consistently.
Phew.
The most confusing part is that all the Python/Jupyter projects currently claim that you don't -- and possibly even shouldn't -- call the commands that `generate-directory` runs. I double-checked, and without them nothing works at all.

**Comment by guaraqe** [comment]
These commands are a result of interaction between Jupyter and Nix - Jupyter's package manager is not determinist, and wants to install things inside the code tree, and Nix wants determinism. That's why we have to generate something outside the Nix store. It's been some time I don't try to tackle this issue again, maybe something changed in the meantime.

**Comment by kirelagin** [comment]
I suppose, this is fixed by #182.
